### PR TITLE
Update tree node to contain interget key and T value, modified method…

### PR DESCRIPTION
**Pull request by ljw7975**
Tree node contains int key and T value,
modified successor method to find_successor,
Add second example
Modified delete method

### CODE REVIEW

File: src/main/java/BinaryTree/BinaryTree.java [hunk 1]
```diff
@@ -14,11 +14,11 @@
...
 
     // Constructors
-    BinaryTree(T key) {
-        current = new BinaryNode<T>(key);
+    BinaryTree(int key, T value) {
+        current = new BinaryNode<T>(key, value);
     }
 
-    public void setLeftChild(T key) {
+    public void setLeftChild(int key) {
```
Comment (ljj7975) [inline src/main/java/BinaryTree/BinaryTree.java]: value?

File: src/main/java/BinaryTree/BinaryTree.java [hunk 2]
```diff
@@ -28,7 +28,7 @@ public void setLeftChild(T key) {
         current.setLeftNode(node);
     }
 
-    public void setRightChild(T key) {
+    public void setRightChild(int key) {
```
Comment (ljj7975) [inline src/main/java/BinaryTree/BinaryTree.java]: value?

File: src/main/java/BinaryTree/AVLtree.java [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+        } else if (node.getLeftNode() != null && node.getRightNode() == null) {
+            node.setHeight(1+current.getLeftNode().getHeight());
+        } else {
+            node.setHeight(1 + Math.max(node.getLeftNode().getHeight(), node.getRightNode().getHeight()));
+        }
+    }
+
+    public int getDepthDifference(BinaryNode<T> target) {
+        if (target.getLeftNode() == null && target.getRightNode() != null) {
+            return (-1)-target.getRightNode().getHeight();
```
Comment (ljj7975) [inline src/main/java/BinaryTree/AVLtree.java]: `-target.getRightNode().getHeight() - 1;` ?

File: src/main/java/BinaryTree/AVLtree.java [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+        } else if (target.getLeftNode() != null && target.getRightNode() == null) {
+            return target.getLeftNode().getHeight()+1;
```
Comment (ljj7975) [inline src/main/java/BinaryTree/AVLtree.java]: `() + 1;`

File: src/main/java/BinaryTree/AVLtree.java [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+        } else if (current.getKey() > insert_key) {
+            if (current.getLeftNode() == null) {
+                setLeftChild(node.getKey(), value);
+            } else {
+                moveToLeftNode();
+                bst_insert(insert_key, value);
+                moveToParentNode();
+            }
+        }
+        updateHeight(current);
```
Comment (ljj7975) [inline src/main/java/BinaryTree/AVLtree.java]: Shouldn't this be propagating upward?

File: src/main/java/BinaryTree/AVLtree.java [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+        } else if (depth_difference < -1 && key < current.getRightNode().getKey()) { // Right Left case
+            moveToRightNode();
+            rightRotation();
+            moveToParentNode();
+            leftRotation();
+        }
+    }
+
+    public void bst_delete(int delete_key) {
+        if (delete_key == 0 || current == null) return;
```
Comment (ljj7975) [inline src/main/java/BinaryTree/AVLtree.java]: what is this `delete_key == 0` case?

File: src/main/java/BinaryTree/AVLtree.java [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+
+        if (current.getKey() < delete_key) {
+            if (current.getRightNode() != null) {
```
Comment (ljj7975) [inline src/main/java/BinaryTree/AVLtree.java]: usually, you are supposed to do
```
if (~~~~) return;
moveToRightNode();
bst_delete(delete_key);
moveToParentNode();
```

File: src/main/java/BinaryTree/AVLtree.java [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+                moveToParentNode();
+            } else {
+                return;
+            }
+        } else if (current.getKey() > delete_key) {
+            moveToLeftNode();
+            bst_delete(delete_key);
+            moveToParentNode();
+        } else if (current.getKey() == delete_key) {
+
```
Comment (ljj7975) [inline src/main/java/BinaryTree/AVLtree.java]: no space

File: src/main/java/BinaryTree/AVLtree.java [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+            // Deleting a node with no children
+            if (current.getLeftNode() == null && current.getRightNode() == null) {
+                if (current.getParentNode().getKey() < current.getKey()) {
+                    current.getParentNode().setRightNode(null);
+                } else {
+                    current.getParentNode().setLeftNode(null);
+                }
+
```
Comment (ljj7975) [inline src/main/java/BinaryTree/AVLtree.java]: no space

File: src/main/java/BinaryTree/AVLtree.java [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+        current = original_position;
+        return successor;
+    }
+
+    public void delete(int delete_key) {
+        // perform standard bst deletion
+        bst_delete(delete_key);
+
+        // balance the tree
+        int depth_difference = getDepthDifference(current);
```
Comment (ljj7975) [inline src/main/java/BinaryTree/AVLtree.java]: create a function `balance_tree_depth` and call it from both insert and delete

File: src/main/java/BinaryTree/AVLtree.java [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+            moveToRightNode();
+            rightRotation();
+            moveToParentNode();
+            leftRotation();
+        }
+    }
+
+    public void rightRotation() {
+        BinaryNode<T> newRoot = current.getLeftNode();
+        BinaryNode<T> moveSubtreeToRight = new BinaryNode<T>();
```
Comment (ljj7975) [inline src/main/java/BinaryTree/AVLtree.java]: why do we need new node?

File: src/main/java/BinaryTree/AVLtree.java [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+        if (newRoot.getRightNode() != null) {
+            moveSubtreeToRight = newRoot.getRightNode();
+        }
+
+        newRoot.setRightNode(current);
+        newRoot.setParentNode(current.getParentNode());
+        current.setParentNode(newRoot);
+        current.setLeftNode(moveSubtreeToRight);
+
+        if (moveSubtreeToRight != null) {
```
Comment (ljj7975) [inline src/main/java/BinaryTree/AVLtree.java]: will this ever be null?

File: src/main/java/BinaryTree/AVLtree.java [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+            moveSubtreeToRight.setParentNode(current);
+        }
+
+        updateHeight(current);
+        updateHeight(newRoot);
+
+        current = newRoot;
+    }
+
+    public void leftRotation() {
```
Comment (ljj7975) [inline src/main/java/BinaryTree/AVLtree.java]: same comment as rightRotation

File: src/main/java/BinaryTree/AVLtreeTest.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+        *           10                                         30
+        *             \                                       /  \
+        *              20                                    20   40
+        *              / \                                  /  \    \
+        *             25  30                              10    25   50
+        *                  \
+        *                   40
+        *                    \
+        *                     50
+        */
```
Comment (ljj7975) [inline src/main/java/BinaryTree/AVLtreeTest.java]: Please add two examples from now on.
As trees are getting more complicated

File: src/main/java/BinaryTree/AVLtree.java (hunk 10 above)
Comment (ljw7975) [reply src/main/java/BinaryTree/AVLtree.java]: insert and deletion have different balancing conditions

File: src/main/java/BinaryTree/AVLtree.java (hunk 12 above)
Comment (ljw7975) [reply src/main/java/BinaryTree/AVLtree.java]: yes in                  5                              3
                        3      7       ->           2      5
                      2                              1           7
                     1
 3 does not have right child to be moved to 5's left node

File: src/main/java/BinaryTree/AVLtree.java (hunk 5 above)
Comment (ljw7975) [reply src/main/java/BinaryTree/AVLtree.java]: what do you mean

File: src/main/java/BinaryTree/AVLtree.java (hunk 12 above)
Comment (ljj7975) [reply src/main/java/BinaryTree/AVLtree.java]: you are setting `moveSubtreeToRight` at line 215.
How can this be not null

**Pull request merged**